crates/topos-core/src/types/stream.rs

```diff
@@ -0,0 +1,133 @@
+use std::{fmt, ops::Deref};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use topos_uci::SubnetId;
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct CertificateSourceStreamPosition {
+    // Source subnet id
+    pub subnet_id: SubnetId,
+    // Source certificate position
+    pub position: Position,
+}
+
+impl fmt::Display for CertificateSourceStreamPosition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.subnet_id, self.position)
+    }
+}
+
+impl CertificateSourceStreamPosition {
+    pub fn new<P: Into<Position>>(subnet_id: SubnetId, position: P) -> Self {
+        Self {
+            subnet_id,
+            position: position.into(),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone, Copy)]
+pub struct CertificateTargetStreamPosition {
+    pub target_subnet_id: SubnetId,
+    pub source_subnet_id: SubnetId,
+    pub position: Position,
+}
+
+impl CertificateTargetStreamPosition {
+    pub fn new<P: Into<Position>>(
+        target_subnet_id: SubnetId,
+        source_subnet_id: SubnetId,
+        position: P,
+    ) -> Self {
+        Self {
+            target_subnet_id,
+            source_subnet_id,
+            position: position.into(),
+        }
+    }
+}
+
+/// Certificate index in the history of the source subnet
```

Comment:
It is used for both source and target subnets yes? Should this be just "…in the history of the subnet."?